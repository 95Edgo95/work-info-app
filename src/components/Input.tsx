import DayPickerInput from "react-day-picker/DayPickerInput";
import classNames from "classnames";
import * as React from "react";

import * as styles from "assets/css/input.css";
import 'react-day-picker/lib/style.css';

function formatDate(date) {
  if (!date) {
    return;
  }

  let value = date;

  if (typeof date === "string") {
    const [month, day, year] = value.replace(new RegExp("[^.0-9]+", "g"), " ").trim().split(" ");

    return `${day}/${month}/${year}`;
  }

  let month = value.getMonth() + 1;
  let day = value.getDate();

  if (day.toString().length === 1) {
    day = `0${day}`;
  }

  if (month.toString().length === 1) {
    month = `0${month}`;
  }

  return `${day}/${month}/${value.getFullYear()}`;
}

interface IInputProps {
  value: string;
  name: string;
  placeholder?: string;
  className?: string;
  disabled?: boolean;
  hasError?: boolean;
  isValid?: boolean;
  error?: string;
  type?: string;

  [key: string]: any;
}

export default function Input(props: IInputProps) {
  const {
    value,
    name,
    isValid = false,
    error = '',
    disabled = false,
    hasError = false,
    placeholder = '',
    className = '',
    type = "text",
    ...otherProps
  } = props;
  const inputWrapperClasses = classNames({
    [styles.inputWrapper]: true,
    [styles.disabled]: disabled,
    [styles.successWrapper]: isValid,
    [styles.errorWrapper]: !!error || hasError,
    [styles.hidden]: type === "hidden",
    [className]: true,
  });

  if (type === "date") {
    const onDayChange = day => (props.onChange && day) ?
      props.onChange({ currentTarget: { name, value: formatDate(day) } }) :
      undefined;

    const onDayEventChange = ({ currentTarget: { value } }) => {
      return props.onChange ?
        props.onChange({ currentTarget: { name, value: formatDate(value) } }) :
        undefined;
    };

    return (
      <React.Fragment>
        <DayPickerInput
          inputProps={{ className: styles.input, name, autoComplete: "off", onChange: onDayEventChange }}
          placeholder={placeholder}
          onDayChange={onDayChange}
          format="DD/MM/YYYY"
          value={value}
        />
        {error && (
          <span className={styles.error}>
          {error}
        </span>
        )}
      </React.Fragment>
    );
  }

  return (
    <div>
      <div className={inputWrapperClasses}>
        <input
          placeholder={placeholder}
          className={styles.input}
          autoComplete="off"
          value={value}
          type={type}
          name={name}
          {...otherProps}
        />
      </div>
      {error && (
        <p className={styles.error}>
          {error}
        </p>
      )}
    </div>
  )
}
