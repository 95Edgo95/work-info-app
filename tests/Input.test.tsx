import Input from "../src/components/Input";
import { mount } from "enzyme";
import * as React from "react";

const changeFn = jest.fn();

describe('Input', () => {
  it('should render correctly', () => {
    const component = mount(
      <Input
        onChange={changeFn}
        value="test"
        name="test"
      />
    );

    expect(component).toMatchSnapshot();
    component.unmount();
  });

  it('should render correctly passport input', () => {
    const component = mount(
      <Input
        onChange={changeFn}
        value="test"
        name="test"
        type="password"
      />
    );

    expect(component).toMatchSnapshot();
    component.unmount();
  });

  it('should render correctly hidden input', () => {
    const component = mount(
      <Input
        onChange={changeFn}
        value="test"
        name="test"
        type="hidden"
      />
    );

    expect(component).toMatchSnapshot();
    component.unmount();
  });

  it('should render correctly with placeholder', () => {
    const component = mount(
      <Input
        onChange={changeFn}
        value="test"
        name="test"
        placeholder="Test"
      />
    );

    expect(component).toMatchSnapshot();
    component.unmount();
  });

  it('should render correctly error input', () => {
    const component = mount(
      <Input
        onChange={changeFn}
        value="test"
        name="test"
        hasError
      />
    );

    expect(component).toMatchSnapshot();
    component.unmount();
  });

  it('should render correctly with error message', () => {
    const component = mount(
      <Input
        onChange={changeFn}
        error="You have error"
        value="test"
        name="test"
      />
    );

    expect(component).toMatchSnapshot();
    component.unmount();
  });

  it('should render correctly disabled input', () => {
    const component = mount(
      <Input
        onChange={changeFn}
        value="test"
        name="test"
        disabled
      />
    );

    expect(component).toMatchSnapshot();
    component.unmount();
  });

  it('should render correctly valid input', () => {
    const component = mount(
      <Input
        onChange={changeFn}
        value="test"
        name="test"
        isValid
      />
    );

    expect(component).toMatchSnapshot();
    component.unmount();
  });
});
