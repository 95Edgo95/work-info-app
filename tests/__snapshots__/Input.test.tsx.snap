// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render correctly 1`] = `
<Input
  name="test"
  onChange={[MockFunction]}
  value="test"
>
  <div>
    <div
      className=""
    >
      <input
        name="test"
        onChange={[MockFunction]}
        placeholder=""
        type="text"
        value="test"
      />
    </div>
  </div>
</Input>
`;

exports[`Input should render correctly disabled input 1`] = `
<Input
  disabled={true}
  name="test"
  onChange={[MockFunction]}
  value="test"
>
  <div>
    <div
      className=""
    >
      <input
        name="test"
        onChange={[MockFunction]}
        placeholder=""
        type="text"
        value="test"
      />
    </div>
  </div>
</Input>
`;

exports[`Input should render correctly error input 1`] = `
<Input
  hasError={true}
  name="test"
  onChange={[MockFunction]}
  value="test"
>
  <div>
    <div
      className=""
    >
      <input
        name="test"
        onChange={[MockFunction]}
        placeholder=""
        type="text"
        value="test"
      />
    </div>
  </div>
</Input>
`;

exports[`Input should render correctly hidden input 1`] = `
<Input
  name="test"
  onChange={[MockFunction]}
  type="hidden"
  value="test"
>
  <div>
    <div
      className="undefined "
    >
      <input
        name="test"
        onChange={[MockFunction]}
        placeholder=""
        type="hidden"
        value="test"
      />
    </div>
  </div>
</Input>
`;

exports[`Input should render correctly passport input 1`] = `
<Input
  name="test"
  onChange={[MockFunction]}
  type="password"
  value="test"
>
  <div>
    <div
      className=""
    >
      <input
        name="test"
        onChange={[MockFunction]}
        placeholder=""
        type="password"
        value="test"
      />
    </div>
  </div>
</Input>
`;

exports[`Input should render correctly valid input 1`] = `
<Input
  isValid={true}
  name="test"
  onChange={[MockFunction]}
  value="test"
>
  <div>
    <div
      className=""
    >
      <input
        name="test"
        onChange={[MockFunction]}
        placeholder=""
        type="text"
        value="test"
      />
    </div>
  </div>
</Input>
`;

exports[`Input should render correctly with error message 1`] = `
<Input
  error="You have error"
  name="test"
  onChange={[MockFunction]}
  value="test"
>
  <div>
    <div
      className=""
    >
      <input
        name="test"
        onChange={[MockFunction]}
        placeholder=""
        type="text"
        value="test"
      />
    </div>
    <p>
      You have error
    </p>
  </div>
</Input>
`;

exports[`Input should render correctly with placeholder 1`] = `
<Input
  name="test"
  onChange={[MockFunction]}
  placeholder="Test"
  value="test"
>
  <div>
    <div
      className=""
    >
      <input
        name="test"
        onChange={[MockFunction]}
        placeholder="Test"
        type="text"
        value="test"
      />
    </div>
  </div>
</Input>
`;
